hyperdrive.go

```diff
@@ -68,31 +79,71 @@ func (hyperdrive *hyperdrive) AcceptTick(t time.Time) {
 		ticks++
 		hyperdrive.ticksPerShard[shardHash] = ticks
 
-		if hyperdrive.ticksPerShard[shardHash] > hyperdrive.timeoutThreshold {
-			// 2. After a number of ticks send a TimedOut transition to the shard
+		if ticks > NumTicksToTriggerTimeOut {
+			// 2. Send a TimedOut transition to the shard
+			if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+				replica.Transition(consensus.TimedOut{Time: t})
+			}
 		}
 	}
 }
 
-func (hyperdrive *hyperdrive) AcceptPropose(shardHash sig.Hash, proposed block.Block) {
-	// TODO:
+func (hyperdrive *hyperdrive) AcceptPropose(shardHash sig.Hash, proposed block.SignedBlock) {
 	// 1. Verify the block is well-formed
-	// 2. Verify the signatory of the block
+	if !hyperdrive.validateBlock(shardHash, proposed) {
+		return
+	}
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.Proposed{SignedBlock: proposed})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptPreVote(shardHash sig.Hash, preVote block.SignedPreVote) {
-	// TODO:
 	// 1. Verify the pre-vote is well-formed
+	if preVote.String() == (block.SignedPreVote{}).String() {
+		return
+	}
+	if preVote.PreVote.Block != nil {
+		if !hyperdrive.validateBlock(shardHash, *preVote.PreVote.Block) {
+			return
+		}
+	}
+	if preVote.PreVote.Round < 0 || preVote.PreVote.Height < 0 {
```

Comment:
We also need to check that the PreVote Round and Height matches the Block Round and Height (if there is a Block). I suggest moving this logic out into a `verifyPreVote` function. Also, is there a reason the verify functions are private? This means they cannot be tested individually.

Reply:
This has been fixed and moved to `ValidatePreVote()` inside `replica/validator.go`